src/WebJobs.Script.WebHost/Controllers/KeysController.cs

```diff
@@ -0,0 +1,158 @@
+
+        [HttpDelete]
+        [Route("admin/functions/{name}/keys/{keyName}")]
+        public IHttpActionResult Delete(string name, string keyName) => DeleteFunctionSecret(keyName, name);
+
+        [HttpDelete]
+        [Route("admin/host/keys/{keyName}")]
+        public IHttpActionResult Delete(string keyName) => DeleteFunctionSecret(keyName);
+
+        private IHttpActionResult GetKeysResult(IDictionary<string, string> keys)
+        {
+            var keysContent = new
+            {
+                keys = keys.Select(k => new { name = k.Key, value = k.Value })
+            };
+
+            var keyResponse = ApiModelUtility.CreateApiModel(keysContent, Request);
+
+            return Ok(keyResponse);
+        }
+
+        private IHttpActionResult PutKey(string keyName, Key key, string functionName = null)
+        {
+            if (functionName != null &&
+                !_scriptHostManager.Instance.Functions.Any(f => string.Equals(f.Name, functionName, StringComparison.OrdinalIgnoreCase)))
+            {
+                return NotFound();
+            }
+
+            if (key?.Value == null)
+            {
+                return BadRequest("Invalid key value");
+            }
+
+            return AddOrUpdateFunctionSecret(keyName, key.Value, functionName);
+        }
+
+        private IHttpActionResult AddOrUpdateFunctionSecret(string keyName, string value, string functionName = null)
+        {
+            if (functionName != null &&
+                !_scriptHostManager.Instance.Functions.Any(f => string.Equals(f.Name, functionName, StringComparison.OrdinalIgnoreCase)))
+            {
+                return NotFound();
+            }
+
+            KeyOperationResult operationResult;
+            if (functionName == null && string.Equals(keyName, MasterKeyName, StringComparison.OrdinalIgnoreCase))
+            {
+                operationResult = _secretManager.SetMasterKey(value);
+            }
+            else
+            {
+                operationResult = _secretManager.AddOrUpdateFunctionSecret(keyName, value, functionName);
+            }
+
+            switch (operationResult.Result)
+            {
+                case OperationResult.Created:
+                    {
+                        var keyResponse = ApiModelUtility.CreateApiModel(new { name = keyName, value = operationResult.Secret }, Request);
+                        return Created(ApiModelUtility.GetBaseUri(Request), keyResponse);
+                    }
+                case OperationResult.Updated:
+                    {
+                        var keyResponse = ApiModelUtility.CreateApiModel(new { name = keyName, value = operationResult.Secret }, Request);
+                        return Ok(keyResponse);
+                    }
+                case OperationResult.NotFound:
+                    return NotFound();
+                case OperationResult.Conflict:
+                    return Conflict();
```

Comment:
Correct, but that is with the current implementation of `ISecretManager`. It is possible that different implementations of `ISecretManager` (or the current, in the future) could return one of those `OperationResult` flags. The controller is just wired to handle them if they happen.
We could simplify this for now, but it does mean it wouldn't handle those scenarios *if* they happened. Didn't want the controller to make assumptions about the `ISecretManager` implementations. 
Thoughts on this?